File: external/lwip/src/sys_arch.zig
// Copyright 2025 Francisco Llobet-Blandino
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// lwIP sys_arch port — board-independent CMSIS-RTOS2 implementation.
//
// This is the root Zig module of the lwip package. It re-exports the lwIP C
// API and provides all sys_arch symbols expected by lwIP when NO_SYS=0.
//
// Board package must inject via addImport/addIncludePath before build:
//   lwip_mod.addImport("cmsis_rtx", cmsis_rtx_mod)
//   lwip_mod.addIncludePath(<board>/lwip_config/)      -- lwipopts.h
//   lwip_mod.addIncludePath(<cmsis_6>/rtos2/include)   -- cmsis_os2.h
//   lwip_mod.addIncludePath(<cmsis_6>/core/include)    -- cmsis_compiler.h
//
// Board provides sys_arch_protect/sys_arch_unprotect (NXP SDK).

const rtx = @import("cmsis_rtx");

// Re-export the lwIP C API so consumers can do:
//   const lwip = @import("lwip");  lwip.c.lwip_init();
pub const c = @import("root.zig").c;

const mu  = rtx.mutex;
const sem = rtx.semaphore;
const mq  = rtx.messageQueue;
const th  = rtx.thread;

const ERR_OK: i8  = 0;
const ERR_MEM: i8 = -1;
const SYS_ARCH_TIMEOUT: u32 = 0xffffffff;
const SYS_MBOX_EMPTY: u32   = 0xffffffff;

// ARM Cortex-M IPSR: 0 = thread mode, non-zero = executing in an ISR.
inline fn getIPSR() u32 {
    return asm volatile ("mrs %[ret], ipsr"
        : [ret] "=r" (-> u32)
    );
}

// Defined by lwIP's sys.c when LWIP_TCPIP_CORE_LOCKING=1.
extern var lock_tcpip_core: mu.osMutexId_t;

// =========================================================================
// Random number generator
// =========================================================================
var rand_value: u32 = 0;

export fn lwip_rand() callconv(.c) u32 {
    rand_value = rand_value *% 1103515245 +% 12345;
    return (rand_value >> 16) % (32767 + 1);
}

// =========================================================================
// Assert
// =========================================================================
export fn sys_assert(pcMessage: [*:0]const u8) callconv(.c) void {
    _ = pcMessage;
    while (true) {}
}

// =========================================================================
// Timing
// =========================================================================
export fn sys_init() callconv(.c) void {}

export fn sys_now() callconv(.c) u32 {
    return rtx.kernel.getTickCount();
}

export fn sys_arch_msleep(delay_ms: u32) callconv(.c) void {
    rtx.osDelay(delay_ms) catch {};
}

// =========================================================================
// Semaphore
// =========================================================================
export fn sys_sem_new(sem_out: *sem.osSemaphoreId_t, count: u8) callconv(.c) i8 {
    sem_out.* = sem.osSemaphoreNew(1, @as(u32, count), null);
    if (sem_out.* == null) return ERR_MEM;
    return ERR_OK;
}

export fn sys_arch_sem_wait(sem_id: *sem.osSemaphoreId_t, timeout_ms: u32) callconv(.c) u32 {
    const t0    = rtx.kernel.getTickCount();
    const ticks = if (timeout_ms == 0) rtx.osWaitForever else timeout_ms;
    if (sem.osSemaphoreAcquire(sem_id.*, ticks) == 0) {
        const elapsed = rtx.kernel.getTickCount() -% t0;
        return if (elapsed == 0) 1 else elapsed;
    }
    return SYS_ARCH_TIMEOUT;
}

export fn sys_sem_signal(sem_id: *sem.osSemaphoreId_t) callconv(.c) void {
    _ = sem.osSemaphoreRelease(sem_id.*);
}

export fn sys_sem_free(sem_id: *sem.osSemaphoreId_t) callconv(.c) void {
    _ = sem.osSemaphoreDelete(sem_id.*);
}

// =========================================================================
// Mutex
// =========================================================================
export fn sys_mutex_new(mutex_out: *mu.osMutexId_t) callconv(.c) i8 {
    const attr = mu.osMutexAttr_t{
        .name     = null,
        .attr_bits = mu.osMutexPrioInherit,
        .cb_mem   = null,
        .cb_size  = 0,
    };
    mutex_out.* = mu.osMutexNew(&attr);
    if (mutex_out.* == null) return ERR_MEM;
    return ERR_OK;
}

export fn sys_mutex_lock(mutex_id: *mu.osMutexId_t) callconv(.c) void {
    _ = mu.osMutexAcquire(mutex_id.*, rtx.osWaitForever);
}

export fn sys_mutex_unlock(mutex_id: *mu.osMutexId_t) callconv(.c) void {
    _ = mu.osMutexRelease(mutex_id.*);
}

export fn sys_mutex_free(mutex_id: *mu.osMutexId_t) callconv(.c) void {
    _ = mu.osMutexDelete(mutex_id.*);
}

// =========================================================================
// Mailbox (message queue carrying void* pointers)
// =========================================================================
export fn sys_mbox_new(mbox_out: *mq.osMessageQueueId_t, size: c_int) callconv(.c) i8 {
    mbox_out.* = mq.osMessageQueueNew(@intCast(size), @sizeOf(?*anyopaque), null);
    if (mbox_out.* == null) return ERR_MEM;
    return ERR_OK;
}

export fn sys_mbox_post(mbox_id: *mq.osMessageQueueId_t, msg: ?*anyopaque) callconv(.c) void {
    var local: ?*anyopaque = msg;
    _ = mq.osMessageQueuePut(mbox_id.*, @as(?*anyopaque, @ptrCast(&local)), 0, rtx.osWaitForever);
}

export fn sys_mbox_trypost(mbox_id: *mq.osMessageQueueId_t, msg: ?*anyopaque) callconv(.c) i8 {
    var local: ?*anyopaque = msg;
    if (mq.osMessageQueuePut(mbox_id.*, @as(?*anyopaque, @ptrCast(&local)), 0, 0) == 0)
        return ERR_OK;
    return ERR_MEM;
}

export fn sys_mbox_trypost_fromisr(mbox_id: *mq.osMessageQueueId_t, msg: ?*anyopaque) callconv(.c) i8 {
    return sys_mbox_trypost(mbox_id, msg);
}

export fn sys_arch_mbox_fetch(mbox_id: *mq.osMessageQueueId_t, msg: ?*?*anyopaque, timeout_ms: u32) callconv(.c) u32 {
    var dummy: ?*anyopaque = null;
    const dst: *?*anyopaque = msg orelse &dummy;
    const t0    = rtx.kernel.getTickCount();
    const ticks = if (timeout_ms == 0) rtx.osWaitForever else timeout_ms;
    if (mq.osMessageQueueGet(mbox_id.*, @as(?*anyopaque, @ptrCast(dst)), null, ticks) == 0) {
        const elapsed = rtx.kernel.getTickCount() -% t0;
        return if (elapsed == 0) 1 else elapsed;
    }
    if (msg) |m| m.* = null;
    return SYS_ARCH_TIMEOUT;
}

export fn sys_arch_mbox_tryfetch(mbox_id: *mq.osMessageQueueId_t, msg: ?*?*anyopaque) callconv(.c) u32 {
    var dummy: ?*anyopaque = null;
    const dst: *?*anyopaque = msg orelse &dummy;
    if (mq.osMessageQueueGet(mbox_id.*, @as(?*anyopaque, @ptrCast(dst)), null, 0) == 0) return 0;
    return SYS_MBOX_EMPTY;
}

export fn sys_mbox_free(mbox_id: *mq.osMessageQueueId_t) callconv(.c) void {
    _ = mq.osMessageQueueDelete(mbox_id.*);
}

// =========================================================================
// Thread
// =========================================================================
export fn sys_thread_new(
    name:       [*:0]const u8,
    thread_fn:  ?*const fn (?*anyopaque) callconv(.c) void,
    arg:        ?*anyopaque,
    stack_size: c_int,
    prio:       c_int,
) callconv(.c) th.osThreadId_t {
    const attr = th.osThreadAttr_t{
        .name         = name,
        .attr_bits    = 0,
        .cb_mem       = null,
        .cb_size      = 0,
        .stack_mem    = null,
        .stack_size   = @intCast(stack_size),
        .priority     = @intFromEnum(th.osThreadPriority.osPriorityNormal) + @as(i32, prio),
        .tz_module    = 0,
        .affinity_mask = 0,
    };
    return th.osThreadNew(@ptrCast(thread_fn), arg, &attr);
}

// =========================================================================
// Core locking (LWIP_TCPIP_CORE_LOCKING=1)
// =========================================================================
var lwip_core_lock_count:  u8              = 0;
var lwip_core_lock_holder: th.osThreadId_t = null;
var lwip_tcpip_thread:     th.osThreadId_t = null;

export fn sys_lock_tcpip_core() callconv(.c) void {
    _ = mu.osMutexAcquire(lock_tcpip_core, rtx.osWaitForever);
    if (lwip_core_lock_count == 0) {
        lwip_core_lock_holder = rtx.thread.getThreadId().id;
    }
    lwip_core_lock_count += 1;
}

export fn sys_unlock_tcpip_core() callconv(.c) void {
    lwip_core_lock_count -= 1;
    if (lwip_core_lock_count == 0) {
        lwip_core_lock_holder = null;
    }
    _ = mu.osMutexRelease(lock_tcpip_core);
}

export fn sys_mark_tcpip_thread() callconv(.c) void {
    lwip_tcpip_thread = rtx.thread.getThreadId().id;
}

export fn sys_check_core_locking() callconv(.c) void {
    if (getIPSR() != 0) {
        while (true) {}
    }
    if (lwip_tcpip_thread != null) {
        const current = rtx.thread.getThreadId().id;
        if (current != lwip_core_lock_holder or lwip_core_lock_count == 0) {
            while (true) {}
        }
    }
}

File: external/lwip/src/sys_arch.c
/*
 * Copyright (c) 2001-2003 Swedish Institute of Computer Science.
 * Copyright (c) 2013-2016, Freescale Semiconductor, Inc.
 * Copyright 2016-2020,2022-2023,2025 NXP
 * All rights reserved.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * lwIP sys_arch — board-independent CMSIS-RTOS2 port.
 *
 * Board-specific functions (sys_arch_protect / sys_arch_unprotect) that call
 * DisableGlobalIRQ / EnableGlobalIRQ from the NXP SDK are provided by the
 * board package (board/ethernet/sys_arch.c), exactly as the board provides
 * RTX_Config.h for the cmsis_rtx package.
 *
 * Consumer must inject include paths for:
 *   - cmsis_os2.h   (CMSIS-RTOS2, from cmsis_6 rtos2/include)
 *   - cmsis_gcc.h   (for __get_IPSR, from cmsis_6 core/include)
 *   - lwipopts.h    (from board's lwip_config/)
 */

#include "arch/sys_arch.h"
#include "cmsis_compiler.h"  /* __get_IPSR — from cmsis_6 core/include, injected by board */
#include "lwip/opt.h"
#include "lwip/debug.h"
#include "lwip/def.h"
#include "lwip/sys.h"
#include "lwip/mem.h"
#include "lwip/stats.h"
#include "lwip/tcpip.h"

/* -------------------------------------------------------------------------
 * Board-independent utilities
 * ------------------------------------------------------------------------- */

static uint32_t _rand_value;
uint32_t lwip_rand(void)
{
    _rand_value = _rand_value * 1103515245u + 12345u;
    return ((uint32_t)(_rand_value >> 16u) % (32767u + 1u));
}

void sys_assert(const char *pcMessage)
{
#ifdef LWIP_DEBUG
    LWIP_PLATFORM_DIAG((pcMessage));
#endif
    for (;;) {}
}

#if !NO_SYS

#include "cmsis_os2.h"

/* -------------------------------------------------------------------------
 * Timing
 * ------------------------------------------------------------------------- */

void sys_init(void) {}

u32_t sys_now(void)
{
    return (u32_t)osKernelGetTickCount();
}

void sys_arch_msleep(u32_t delay_ms)
{
    osDelay((uint32_t)delay_ms);
}

/* -------------------------------------------------------------------------
 * Semaphore
 * ------------------------------------------------------------------------- */

err_t sys_sem_new(sys_sem_t *sem, u8_t count)
{
    LWIP_ASSERT("sem != NULL", sem != NULL);
    *sem = osSemaphoreNew(1, (uint32_t)count, NULL);
    if (*sem == NULL) {
        SYS_STATS_INC(sem.err);
        return ERR_MEM;
    }
    SYS_STATS_INC_USED(sem);
    return ERR_OK;
}

u32_t sys_arch_sem_wait(sys_sem_t *sem, u32_t timeout_ms)
{
    uint32_t t0    = osKernelGetTickCount();
    uint32_t ticks = (timeout_ms == 0u) ? osWaitForever : (uint32_t)timeout_ms;
    osStatus_t st  = osSemaphoreAcquire(*sem, ticks);
    if (st == osOK) {
        uint32_t elapsed = osKernelGetTickCount() - t0;
        return (elapsed == 0u) ? 1u : elapsed;
    }
    return SYS_ARCH_TIMEOUT;
}

void sys_sem_signal(sys_sem_t *sem)
{
    osSemaphoreRelease(*sem);
}

void sys_sem_free(sys_sem_t *sem)
{
    SYS_STATS_DEC(sem.used);
    osSemaphoreDelete(*sem);
}

/* -------------------------------------------------------------------------
 * Mutex
 * ------------------------------------------------------------------------- */

err_t sys_mutex_new(sys_mutex_t *mutex)
{
    static const osMutexAttr_t attr = { .attr_bits = osMutexPrioInherit };
    *mutex = osMutexNew(&attr);
    if (*mutex == NULL) {
        SYS_STATS_INC(mutex.err);
        return ERR_MEM;
    }
    SYS_STATS_INC_USED(mutex);
    return ERR_OK;
}

void sys_mutex_lock(sys_mutex_t *mutex)
{
    osMutexAcquire(*mutex, osWaitForever);
}

void sys_mutex_unlock(sys_mutex_t *mutex)
{
    osMutexRelease(*mutex);
}

void sys_mutex_free(sys_mutex_t *mutex)
{
    SYS_STATS_DEC(mutex.used);
    osMutexDelete(*mutex);
}

/* -------------------------------------------------------------------------
 * Mailbox
 * ------------------------------------------------------------------------- */

err_t sys_mbox_new(sys_mbox_t *mbox, int size)
{
    *mbox = osMessageQueueNew((uint32_t)size, sizeof(void *), NULL);
    if (*mbox == NULL) {
        SYS_STATS_INC(mbox.err);
        return ERR_MEM;
    }
    SYS_STATS_INC_USED(mbox);
    return ERR_OK;
}

void sys_mbox_post(sys_mbox_t *mbox, void *msg)
{
    osMessageQueuePut(*mbox, &msg, 0u, osWaitForever);
}

err_t sys_mbox_trypost(sys_mbox_t *mbox, void *msg)
{
    if (osMessageQueuePut(*mbox, &msg, 0u, 0u) == osOK) {
        return ERR_OK;
    }
    SYS_STATS_INC(mbox.err);
    return ERR_MEM;
}

err_t sys_mbox_trypost_fromisr(sys_mbox_t *mbox, void *msg)
{
    return sys_mbox_trypost(mbox, msg);
}

u32_t sys_arch_mbox_fetch(sys_mbox_t *mbox, void **msg, u32_t timeout_ms)
{
    void  *dummy;
    if (msg == NULL) {
        msg = &dummy;
    }
    uint32_t t0    = osKernelGetTickCount();
    uint32_t ticks = (timeout_ms == 0u) ? osWaitForever : (uint32_t)timeout_ms;
    osStatus_t st  = osMessageQueueGet(*mbox, msg, NULL, ticks);
    if (st == osOK) {
        uint32_t elapsed = osKernelGetTickCount() - t0;
        return (elapsed == 0u) ? 1u : elapsed;
    }
    *msg = NULL;
    return SYS_ARCH_TIMEOUT;
}

u32_t sys_arch_mbox_tryfetch(sys_mbox_t *mbox, void **msg)
{
    void *dummy;
    if (msg == NULL) {
        msg = &dummy;
    }
    return (osMessageQueueGet(*mbox, msg, NULL, 0u) == osOK) ? 0u : (u32_t)SYS_MBOX_EMPTY;
}

void sys_mbox_free(sys_mbox_t *mbox)
{
    SYS_STATS_DEC(mbox.used);
    osMessageQueueDelete(*mbox);
}

/* -------------------------------------------------------------------------
 * Thread
 * ------------------------------------------------------------------------- */

sys_thread_t sys_thread_new(const char *name, void (*fn)(void *), void *arg,
                             int stack_size, int prio)
{
    osThreadAttr_t attr = {
        .name       = name,
        .stack_size = (uint32_t)stack_size,
        .priority   = (osPriority_t)(osPriorityNormal + prio),
    };
    osThreadId_t id = osThreadNew(fn, arg, &attr);
    LWIP_ASSERT("sys_thread_new: osThreadNew failed", id != NULL);
    return id;
}

/* -------------------------------------------------------------------------
 * Core locking
 * ------------------------------------------------------------------------- */

#if LWIP_TCPIP_CORE_LOCKING

static u8_t          lwip_core_lock_count;
static osThreadId_t  lwip_core_lock_holder;

void sys_lock_tcpip_core(void)
{
    sys_mutex_lock(&lock_tcpip_core);
    if (lwip_core_lock_count == 0u) {
        lwip_core_lock_holder = osThreadGetId();
    }
    lwip_core_lock_count++;
}

void sys_unlock_tcpip_core(void)
{
    lwip_core_lock_count--;
    if (lwip_core_lock_count == 0u) {
        lwip_core_lock_holder = NULL;
    }
    sys_mutex_unlock(&lock_tcpip_core);
}

#endif /* LWIP_TCPIP_CORE_LOCKING */

static osThreadId_t lwip_tcpip_thread;

void sys_mark_tcpip_thread(void)
{
    lwip_tcpip_thread = osThreadGetId();
}

void sys_check_core_locking(void)
{
    LWIP_ASSERT("Function called from interrupt context", (__get_IPSR() == 0));

    if (lwip_tcpip_thread != NULL) {
        osThreadId_t current = osThreadGetId();
        LWIP_UNUSED_ARG(current);
#if LWIP_TCPIP_CORE_LOCKING
        LWIP_UNUSED_ARG(lwip_core_lock_holder);
        LWIP_ASSERT("Function called without core lock",
                    current == lwip_core_lock_holder && lwip_core_lock_count > 0u);
#else
        LWIP_ASSERT("Function called from wrong thread", current == lwip_tcpip_thread);
#endif
    }
}

#endif /* !NO_SYS */
